src/Language/PureScript/Sugar/ObjectWildcards.hs

```diff
@@ -3,19 +3,48 @@ module Language.PureScript.Sugar.ObjectWildcards
   , desugarDecl
   ) where
 
-import Prelude.Compat
+import           Prelude.Compat
 
-import Control.Monad (forM)
-import Control.Monad.Error.Class (MonadError(..))
-import Control.Monad.Supply.Class
+import           Control.Monad (forM, foldM)
+import           Control.Monad.Error.Class (MonadError(..))
+import           Control.Monad.Supply.Class
+import           Data.List (partition)
+import           Data.List.NonEmpty (NonEmpty(..))
+import           Data.Map (Map)
+import qualified Data.Map as M
+import           Data.Maybe (catMaybes)
+import           Language.PureScript.AST
+import           Language.PureScript.Environment (NameKind(..))
+import           Language.PureScript.Errors
+import           Language.PureScript.Label (Label(..))
+import           Language.PureScript.Names
+import           Language.PureScript.PSString (PSString)
 
-import Data.List (partition)
-import Data.Maybe (catMaybes)
-
-import Language.PureScript.AST
-import Language.PureScript.Errors
-import Language.PureScript.Names
-import Language.PureScript.PSString (PSString)
+-- `PathNode` and `PathTree` are used as an intermediate form when desugaring a nested object update.
+-- For an update such as:
+--
+--  x { foo = 0
+--    , bar.baz = 1
+--    , bar.qux = 2 }
+--
+-- We represent the updates as the `PathTree`:
+--
+--  M.fromList [ ("foo", Leaf 3)
+--             , ("bar", Branch (M.fromList [ ("baz", Leaf 1)
+--                                          , ("qux", Leaf 2) ]) ])
+--
+-- Which we then convert to an expression representing the following:
+--
+--   let x' = x
+--   in x' { foo = 0
+--         , bar = x'.bar { baz = 1
+--                        , qux = 2 } }
+--
+-- The `let` here is required to prevent re-evaluating the object expression `x`.
+-- However we don't generate this when using an anonymous argument for the object.
```

Comment:
With regular object updates, in our AST, the object expression is only mentioned once.  In the conversion to JavaScript, the expression is evaluated and its result is cached in a variable when building up the new object - but this is a JavaScript implementation detail.

In this case we would be duplicating the expression over multiple unrelated object updates - each of which would evaluate it separately.

```purescript
(f x) { foo.bar = 2 }

-- without a let, this would evaluate `f x` multiple times
(f x) { foo = (f x).foo { bar = 2 } }

-- with the let, we don't
let y = f x in y { foo = y.foo { bar = 2 } }
```